editor/run_instances_dialog.cpp

```diff
@@ -0,0 +1,289 @@
+			// Append Godot-specific custom arguments.
+			custom_args = _split_cmdline_args(raw_custom_args);
+			for (int i = 0; i < custom_args.size(); i++) {
+				r_list.push_back(custom_args[i].replace(" ", "%20"));
+			}
+		}
+	}
+
+	if (use_multiple_instances && !override_args) {
+		r_list.push_back(argument_list[p_idx]);
+	}
+}
+
+RunInstancesDialog::RunInstancesDialog() {
+	singleton = this;
+
+	main_apply_timer = memnew(Timer);
+	add_child(main_apply_timer);
+	main_apply_timer->set_wait_time(0.5);
+	main_apply_timer->set_one_shot(true);
+	main_apply_timer->connect("timeout", callable_mp(this, &RunInstancesDialog::_save_main_args));
+
+	instance_apply_timer = memnew(Timer);
+	add_child(instance_apply_timer);
+	instance_apply_timer->set_wait_time(0.5);
+	instance_apply_timer->set_one_shot(true);
+	instance_apply_timer->connect("timeout", callable_mp(this, &RunInstancesDialog::_save_arguments));
+
+	VBoxContainer *main_vb = memnew(VBoxContainer);
+	add_child(main_vb);
+
+	{
+		Label *l = memnew(Label);
+		main_vb->add_child(l);
+		l->set_text(TTR("Main Run Args:"));
+	}
+
+	main_args_edit = memnew(LineEdit);
+	main_vb->add_child(main_args_edit);
+	_fetch_main_args();
+	ProjectSettings::get_singleton()->connect("settings_changed", callable_mp(this, &RunInstancesDialog::_fetch_main_args));
+	main_args_edit->connect("text_changed", callable_mp(this, &RunInstancesDialog::_start_main_timer).unbind(1));
+
+	enable_multiple_instances_checkbox = memnew(CheckBox);
+	main_vb->add_child(enable_multiple_instances_checkbox);
+	enable_multiple_instances_checkbox->set_text(TTR("Enable Multiple Instances"));
+	enable_multiple_instances_checkbox->set_pressed(EditorSettings::get_singleton()->get_project_metadata("debug_options", "multiple_instances_enabled", false));
+	enable_multiple_instances_checkbox->connect("pressed", callable_mp(this, &RunInstancesDialog::_start_instance_timer));
+
+	override_list = EditorSettings::get_singleton()->get_project_metadata("debug_options", "multiple_instances_overrides", varray(false, false, false, false));
+	argument_list = EditorSettings::get_singleton()->get_project_metadata("debug_options", "multiple_instances_arguments", PackedStringArray{ "", "", "", "" });
+
+	instance_count = memnew(SpinBox);
+	main_vb->add_child(instance_count);
+	instance_count->set_min(1);
+	instance_count->set_max(12);
+	instance_count->set_value(argument_list.size());
+	instance_count->connect("value_changed", callable_mp(this, &RunInstancesDialog::_start_instance_timer).unbind(1));
+	instance_count->connect("value_changed", callable_mp(this, &RunInstancesDialog::_refresh_argument_count).unbind(1));
+	enable_multiple_instances_checkbox->connect("toggled", callable_mp(instance_count, &SpinBox::set_editable));
+
+	Label *l = memnew(Label);
+	main_vb->add_child(l);
+	l->set_text(TTR("Launch Arguments"));
+	l->set_h_size_flags(Control::SIZE_SHRINK_CENTER);
+
+	argument_container = memnew(VBoxContainer);
+	main_vb->add_child(argument_container);
```

Comment:
There should be a `ScrollContainer` containing the `argument_container`, in order to make sure that it doesn't add children indefinitely.

Reply:
That's what instance limit of 12 is for. I didn't want to use ScrollContainer, because it makes the size undefined, but I guess I could just give it a minimum size 🤔
I will improve this in the follow-up, for now I think it's fine as is.